Description: Add checks to prevent crashing of snespurify when scanning broken zip files.
Origin: http://board.byuu.org/viewtopic.php?p=44755#p44755

Index: bsnes-090/bsnes/nall/zip.hpp
===================================================================
--- bsnes-090.orig/bsnes/nall/zip.hpp	2012-08-06 22:51:07.000000000 +0200
+++ bsnes-090/bsnes/nall/zip.hpp	2012-08-07 18:11:39.351685329 +0200
@@ -47,7 +47,7 @@
     }
     const uint8_t *directory = data + read(footer + 16, 4);
 
-    while(true) {
+    while(directory < footer - 46) {
       unsigned signature = read(directory + 0, 4);
       if(signature != 0x02014b50) break;
 
@@ -56,10 +56,12 @@
       file.crc32 = read(directory + 16, 4);
       file.csize = read(directory + 20, 4);
       file.size  = read(directory + 24, 4);
+      if (file.csize > size) break;
 
       unsigned namelength = read(directory + 28, 2);
       unsigned extralength = read(directory + 30, 2);
       unsigned commentlength = read(directory + 32, 2);
+      if(directory > footer - 46 - namelength) break;
 
       char *filename = new char[namelength + 1];
       memcpy(filename, directory + 46, namelength);
@@ -68,9 +70,11 @@
       delete[] filename;
 
       unsigned offset = read(directory + 42, 4);
+      if(offset > size - 30) break;
       unsigned offsetNL = read(data + offset + 26, 2);
       unsigned offsetEL = read(data + offset + 28, 2);
       file.data = data + offset + 30 + offsetNL + offsetEL;
+      if(file.data > footer - file.csize) break;
 
       directory += 46 + namelength + extralength + commentlength;
 
Index: bsnes-090/kaijuu/nall/zip.hpp
===================================================================
--- bsnes-090.orig/kaijuu/nall/zip.hpp	2012-07-29 15:08:31.000000000 +0200
+++ bsnes-090/kaijuu/nall/zip.hpp	2012-08-07 18:11:38.315685338 +0200
@@ -47,7 +47,7 @@
     }
     const uint8_t *directory = data + read(footer + 16, 4);
 
-    while(true) {
+    while(directory < footer - 46) {
       unsigned signature = read(directory + 0, 4);
       if(signature != 0x02014b50) break;
 
@@ -56,10 +56,12 @@
       file.crc32 = read(directory + 16, 4);
       file.csize = read(directory + 20, 4);
       file.size  = read(directory + 24, 4);
+      if (file.csize > size) break;
 
       unsigned namelength = read(directory + 28, 2);
       unsigned extralength = read(directory + 30, 2);
       unsigned commentlength = read(directory + 32, 2);
+      if(directory > footer - 46 - namelength) break;
 
       char *filename = new char[namelength + 1];
       memcpy(filename, directory + 46, namelength);
@@ -68,9 +70,11 @@
       delete[] filename;
 
       unsigned offset = read(directory + 42, 4);
+      if(offset > size - 30) break;
       unsigned offsetNL = read(data + offset + 26, 2);
       unsigned offsetEL = read(data + offset + 28, 2);
       file.data = data + offset + 30 + offsetNL + offsetEL;
+      if(file.data > footer - file.csize) break;
 
       directory += 46 + namelength + extralength + commentlength;
 
Index: bsnes-090/purify/nall/zip.hpp
===================================================================
--- bsnes-090.orig/purify/nall/zip.hpp	2012-08-06 22:48:58.000000000 +0200
+++ bsnes-090/purify/nall/zip.hpp	2012-08-07 18:11:36.631685353 +0200
@@ -47,7 +47,7 @@
     }
     const uint8_t *directory = data + read(footer + 16, 4);
 
-    while(true) {
+    while(directory < footer - 46) {
       unsigned signature = read(directory + 0, 4);
       if(signature != 0x02014b50) break;
 
@@ -56,10 +56,12 @@
       file.crc32 = read(directory + 16, 4);
       file.csize = read(directory + 20, 4);
       file.size  = read(directory + 24, 4);
+      if (file.csize > size) break;
 
       unsigned namelength = read(directory + 28, 2);
       unsigned extralength = read(directory + 30, 2);
       unsigned commentlength = read(directory + 32, 2);
+      if(directory > footer - 46 - namelength) break;
 
       char *filename = new char[namelength + 1];
       memcpy(filename, directory + 46, namelength);
@@ -68,9 +70,11 @@
       delete[] filename;
 
       unsigned offset = read(directory + 42, 4);
+      if(offset > size - 30) break;
       unsigned offsetNL = read(data + offset + 26, 2);
       unsigned offsetEL = read(data + offset + 28, 2);
       file.data = data + offset + 30 + offsetNL + offsetEL;
+      if(file.data > footer - file.csize) break;
 
       directory += 46 + namelength + extralength + commentlength;
 
